---
title: "Delta Panel Structure Notebook"
author: "Katherine Silliman"
date: "12/15/21"
output:
  html_document:
    df_print: paged
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook for running structure from a MassArray Genotypes output, specific for the 39/52 plex Delta Bass panel. To execute code chunks, click the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*.

# Installing packages (ONLY RUN ONCE)
If this is the 1st time you are using this notebook, you need to install these R packages. R version > 4.0 is also required.  
```{r, eval=FALSE}
if(!require(devtools)) install.packages("devtools")
library(devtools)
install_github("ksil91/strataG")
install.packages("reticulate")
```
## Python install
You also need Python 3 installed on your system, and then following Python packages: `pandas` and `xlrd`. A tutorial for installing Python packages with `pip` can be found [here](https://packaging.python.org/tutorials/installing-packages/). 

In Terminal, you will type:
```
python3 -m pip install pandas
python3 -m pip install xlrd
```

# Read in data (SET FILE NAMES EVERY TIME)
Now load those libraries:
```{r}
library(StrStrataGKS)
library(reticulate)
```

Now we take the Excel GenotypeArea output from the MassArray software, read it into R, and remove any markers we want to remove. If you want to remove individuals from the Structure analysis or combine individuals from multiple runs into 1 excel file, you can do that in the "Genotypes" sheet, but otherwise do NOT edit this file or change the column names.  

### NOTE
If there is a marker with absolutely no data for any samples (usually a known bad marker), you need to add data for at least 1 sample in the Excel file or there will be an error. Data can be anything (eg, AT), since you will remove this marker before running structure. 

If you have duplicate sample names n the Excel file, you need to change them so they are unique (eg, add a B to the end of one).

Look to see where the Genotypes Excel file is located on your computer, and specify that file location (path) and name in the code chunk below. If the file is called "GenotypeArea_Megan_38FLNB_CHIP979_10222021_CHIP979_10222021_1_.xls" and is on your Desktop, it would look like this: `"~/Desktop/GenotypeArea_Megan_38FLNB_CHIP979_10222021_CHIP979_10222021_1_.xls"`. 



You also need to specify what your output file will be titled. This should be specific to the project:
```{r}
#change these EVERY TIME
inputFile1 = "GenotypeArea_Katherine_DB_CHIP834_DB52_12072020_CHIP834_DB52_12072020_1_ACme.xls"
inputFile2 = "GenotypeArea_Katherine_DB_CHIP835_DB39_12072020_CHIP835_D392_12072020_1_ACme.xls"
outputFile = "delta_test"
```
Now double check that the file below is the reference file you want to use. This is a CSV file specific to the panel being used, and contains genotypes for all markers (even those typically excluded). For the Delta panel, the file is `refs_Delta_91.csv`:
```{r}
refFile = "refs_Delta_91.csv"
```

Now run this Python code to convert the Excel file into a CSV file for reading into R. Written for Python3. 
```{python, results = 'hide'}
import pandas as pd

infile1 = r.inputFile1
infile2 = r.inputFile2

inxl1 = pd.ExcelFile(infile1)
inxl1 = inxl1.parse("Genotypes")
inxl1.set_index("SAMPLE_NAME", drop = True, inplace=True)
if "WELL" in inxl1.columns.values:
  inxl1 = inxl1.drop(columns=["WELL"])

## if you added an X before the sample names on the 2nd panel, uncomment below
#def stripX(x):
#  return x[1:]
#inxl2 = pd.read_excel(infile2, sheetname="Genotypes", converters={'SAMPLE_NAME':stripX})

## run if no X added to sample names
inxl2 = pd.ExcelFile(infile2)

inxl2 = inxl2.parse("Genotypes")
inxl2.set_index("SAMPLE_NAME", drop = True, inplace=True)
if "WELL" in inxl2.columns.values:
  inxl2 = inxl2.drop(columns=["WELL"])

#combine sheets
comb = pd.concat([inxl1,inxl2],axis=1)

#convert homozygotes to 2 alleles
alleles = {'A':'AA','T':'TT','C':'CC','G':'GG'}
comb2 = comb.replace(to_replace=alleles)
[comb2.insert(i, comb2.columns.values[i]+".1", comb2.iloc[:,i].str[1]) for i in range(comb2.shape[1]-1,-1,-1)]
for i in range(1,comb2.shape[1],2):
    comb2.iloc[:,i] = pd.array(comb2.iloc[:,i].str[0])
    comb2.columns.values[i] = comb2.columns.values[i]+".2"

```

Make sure that the output has all your samples as row names, loci as column names, and 2 columns for each marker (182 for Delta panel).
```{python}
comb2.head()
# write output to a new file, to be read into by R
comb2.to_csv(r.outputFile+"_df.csv",index_label="ids")
```
The Python code created a new csv file in the location specified by `outputFile`. This will be read into by R with the StrStrataGKS package. 
```{r}
gen.data <- readGenData(paste0(outputFile,"_df.csv"))
# create a vector of "unknowns" the same length as the number of samples
Strata <- c(rep("unknown",nrow(gen.data)))
gen.merge <- cbind(Strata, gen.data)

#read in reference, which already has strata info
ref.data <- readGenData(refFile)

# create a single data.frame with ref and data
all.data <- rbind(ref.data,gen.merge)
```

# Drop bad markers  
You need to specify which markers needed to be dropped before running structure. These are the 18 markers dropped for the Delta panel to get to 73 SNPs. These are listed in a text file, and can be edited there if needed (you need to list each marker 2x, once with .1 and once with .2):
```{r}
# CHANGE if using different marker file
badFile <- "delta_badSNPs"
# Read in the data
x <- scan(badFile, what="", sep="\n")
names.use <- names(all.data)[!(names(all.data) %in% x)]
all.data <- all.data[,names.use]
```

# Convert to a gtype object, then run Structure
```{r}
#convert to gtypes object
all.gtype <- df2gtypes(all.data, ploidy = 2, id.col = 1, strata.col = 2)

# make popflag vector, specifying reference individuals with a 1 and unknowns with a 0
pflag <- c(rep(1,length(getIndNames(all.gtype))))
pflag[grepl("unknown",as.vector(getStrata(all.gtype)))] <- 0
```

Run structure, specifying the number of Ks and replicate runs. For the Delta panel, K=3 and we want 3 replicate runs. Always do `inferalpha=TRUE`
```{r results='hide'}
K <- 3
R <- 3
sr <- structureRun(all.gtype, label = outputFile,k.range = K, num.k.rep = R,
                   burnin=20000,numreps=150000,noadmix=F, freqscorr=T,pop.prior = "usepopinfo",
                   migrprior = 0.05,popflag=pflag,inferalpha=T,delete.files = FALSE)
```

Now average across all 3 runs and save to a .csv file that can be opened in Excel.
```{r}
qmat <- clumpp(sr,k=K)
write.csv(qmat,row.names = F,file = paste0(outputFile,"_qvalues.csv"))
#if you want to plot the results, uncomment below
#structurePlot(qmat)
```

# Creating a reference file (ONLY IF YOU NEED A NEW REFERENCE FILE) 
We need to read in the Excel files of reference genotypes for both the 52 snp and 39 snp panels, format them, then save them to a CSV to be used with structure in R.

```{r}
refExcel1 = "refs_Delta_52_1121.xlsx"
refExcel2 = "refs_Delta_39_1121.xlsx"
refOutput = "refs_Delta_91.csv"
```

```{python, results='hide'}
import pandas as pd

infile1 = r.refExcel1
infile2 = r.refExcel2

inxl1 = pd.ExcelFile(infile1)
inxl1 = inxl1.parse("Genotypes")
inxl1.set_index("SAMPLE_NAME", drop = True, inplace=True)
if "WELL" in inxl1.columns.values:
  inxl1 = inxl1.drop(columns=["WELL"])

inxl1= inxl1.drop(columns=["Strata"])
## if you added an X before the sample names on the 2nd panel, uncomment below
#def stripX(x):
#  return x[1:]
#inxl2 = pd.read_excel(infile2, sheetname="Genotypes", converters={'SAMPLE_NAME':stripX})

## run if no X added to sample names
inxl2 = pd.ExcelFile(infile2)

inxl2 = inxl2.parse("Genotypes")
inxl2.set_index("SAMPLE_NAME", drop = True, inplace=True)
if "WELL" in inxl2.columns.values:
  inxl2 = inxl2.drop(columns=["WELL"])

comb = pd.concat([inxl1,inxl2],axis=1)

# drop columns with strata/species info, to be added back later
strata = comb[["Strata"]]
comb= comb.drop(columns=["Strata"])
#convert homozygotes to 2 alleles
alleles = {'A':'AA','T':'TT','C':'CC','G':'GG'}
comb2 = comb.replace(to_replace=alleles)
[comb2.insert(i, comb2.columns.values[i]+".1", comb2.iloc[:,i].str[1]) for i in range(comb2.shape[1]-1,-1,-1)]
for i in range(1,comb2.shape[1],2):
    comb2.iloc[:,i] = pd.array(comb2.iloc[:,i].str[0])
    comb2.columns.values[i] = comb2.columns.values[i]+".2"
combMerge = strata.join(comb2)

#add rf to every sample name to avoid duplicate names
combMerge.rename(index=lambda x: x+"rf,", inplace=True)
```
Write formatted dataframe to CSV file. This is what you'll read in for a Structure analysis.
```{python}
combMerge.to_csv(r.refOutput,index_label="ids")
```


